add blog about stacks in the JavaScript backend

luite opened the pull request:
This is the updated/expanded blog post about stacks for the JS backend, that introduces the trapolining calling convention in a number of steps, and explains `h$stack` based on that. Exceptions and threads are now at the end.

hsyl20 reviewed (commented):
Good writing!

Could you add a short section about `throwTo` (`killThread#`)? I think it could be a good opportunity to document how it works: a thread pushing a stack frame on another thread's stack, or in its `excep` "stack" if the target is currently masking exceptions. And then the check in the scheduler loop for pending async exceptions.

hsyl20 commented on 2023-04-21-js-backend-stacks.md:
```diff
@@ -0,0 +1,459 @@
+| Slot | Value                                  | Frame | Type / Description |
+| ---- | -------------------------------------- | ----- | ----------------- |
+| 0    | `h$done`                               | __1__ | `function`, "finished" frame |
+| 1    | `0`                                    | 2     | `number`, fags |
```
```suggestion
| 1    | `0`                                    | 2     | `number`, flags |
```

luite commented:
> Good writing!
> 
> Could you add a short section about `throwTo` (`killThread#`)? I think it could be a good opportunity to document how it works: a thread pushing a stack frame on another thread's stack, or in its `excep` "stack" if the target is currently masking exceptions. And then the check in the scheduler loop for pending async exceptions.

I've applied your other suggestions.

As for async exceptions, I'm not sure where to put it to match the "flow" of the post. I wanted to do exceptions first, since it naturally follows from metadata. But at the point where exceptions are introduced I haven't mentioned yet that threads have their own lightweight stack and thread state object.

It seems to be a more advanced topic

hsyl20 commented:
> > Good writing!
> > Could you add a short section about `throwTo` (`killThread#`)? I think it could be a good opportunity to document how it works: a thread pushing a stack frame on another thread's stack, or in its `excep` "stack" if the target is currently masking exceptions. And then the check in the scheduler loop for pending async exceptions.
> 
> I've applied your other suggestions.
> 
> As for async exceptions, I'm not sure where to put it to match the "flow" of the post. I wanted to do exceptions first, since it naturally follows from metadata. But at the point where exceptions are introduced I haven't mentioned yet that threads have their own lightweight stack and thread state object.
> 
> It seems to be a more advanced topic

I would put it just before the "Important Values and Properties" title. You've introduced threads so it's a good place to say something like:

> ### Thread and asynchronous exceptions
>
> Hey btw a thread can also raise an exception in another thread. Because the runtime is single threaded we don't have to worry about the target thread being executed concurrently, so it's quite simple: if the target thread isn't the current thread, we only have to push a stack frame onto the target thread's stack. The next time the target thread will be scheduled, it will resume its execution with the stack frame we've just pushed which simply raises an exception.
>
> The only thing that makes this more complicated is exception masking: a thread can opt in masking asynchronous exceptions temporarily. While it is in this mode, other threads aren't allowed to push asynchronous exception frames on its stack. What they do instead is to push exceptions on another thread-private stack dedicated to this (see `h$Thread.excep`). These exceptions are then only raised when a thread leaves its exception-masking state.

luite commented:
> > ### Thread and asynchronous exceptions
> > Hey btw a thread can also raise an exception in another thread. Because the runtime is single threaded we don't have to worry about the target thread being executed concurrently, so it's quite simple: if the target thread isn't the current thread, we only have to push a stack frame onto the target thread's stack. The next time the target thread will be scheduled, it will resume its execution with the stack frame we've just pushed which simply raises an exception.
> > The only thing that makes this more complicated is exception masking: a thread can opt in masking asynchronous exceptions temporarily. While it is in this mode, other threads aren't allowed to push asynchronous exception frames on its stack. What they do instead is to push exceptions on another thread-private stack dedicated to this (see `h$Thread.excep`). These exceptions are then only raised when a thread leaves its exception-masking state.

Okay I'll try to word that to fit in the rest of the post.

There are more complications though: asynchronous exceptions require dealing with black holes differently, there's a whole infrastructure to "heapify" part of the stack and stackify it later

luite commented:
> Okay I'll try to word that to fit in the rest of the post.
> 
> There are more complications though: asynchronous exceptions require dealing with black holes differently, there's a whole infrastructure to "heapify" part of the stack and stackify it later

Added the section, including a table of highlighting the operational differences between synchronous and asynchronous threads and a table describing the additional properties of `h$Thread` required (and I added `h$Thread.excep` to the constructor example code above for completion).

Hopefully that's enough depth for this topic

hsyl20 commented:
LGTM, you can publish tomorrow

luite commented:
Okay I'll read through it one more time tomorrow after some rest and publish.

hsyl20 commented:
\o/ Don't forget to promote it on reddit/discourse
